File: SpaceInvaders/GameObject/Alien/Alien.cs
using System;
using System.Diagnostics;

namespace SpaceInvaders
{
    public abstract class Alien : Leaf
    {

        //----------------------------------------------------------------------
        // Constructors / Destructor
        //----------------------------------------------------------------------

        protected Alien(GameObject.Name gameName, Sprite.Name spriteName)
            : base(gameName, spriteName)
        {
            // Do nothing at this layer
        }

        ~Alien()
        {
            // Do nothing at this layer
        }

        //----------------------------------------------------------------------
        // Abstract Methods
        //----------------------------------------------------------------------
        // Strategy
        public abstract int GetPoints();

        //----------------------------------------------------------------------
        // Methods
        //----------------------------------------------------------------------
        public override void Accept(CollisionVisitor other)
        {
            other.VisitAlien(this);
        }

        override public void Move(float xDelta, float yDelta)
        {
            this.x = this.x + xDelta;
            this.y = this.y + yDelta;

        }
        
        public override void VisitShieldBrick(ShieldBrick sb)
        {
            //ShieldBrick vs Missile
            Debug.WriteLine("         collide:  {0} <-> {1}", sb.GetName(), this.name);

            // Notify Observers
            CollisionPair pColPair = CollisionPairManager.GetActiveColPair();
            Debug.Assert(pColPair != null);
            pColPair.SetCollision(sb, this);
            pColPair.NotifyListeners();
        }

        public enum Type
        {
            Squid,
            Crab,
            JellyFish,
            FlyingSaucer
        }
    }
}

File: quarkus-migration/space-invaders-quarkus/src/main/java/net/beeland/spaceinvaders/gameobject/Alien.java
package net.beeland.spaceinvaders.gameobject;

/**
 * Alien - Abstract base class for all alien types
 * Aliens move in formation and can be destroyed for points
 */
public abstract class Alien extends Leaf {
    
    /**
     * Alien type enumeration
     */
    public enum Type {
        SQUID,
        CRAB,
        JELLYFISH,
        FLYING_SAUCER
    }
    
    /**
     * Constructor
     *
     * @param name GameObject name
     */
    protected Alien(GameObject.GameObjectName name) {
        super(name);
    }
    
    @Override
    public void move(float deltaX, float deltaY) {
        this.x += deltaX;
        this.y += deltaY;
    }
    
    @Override
    public void accept(GameObject.CollisionVisitor other) {
        other.visit(this);
    }
    
    /**
     * Get the point value for destroying this alien
     * 
     * @return Point value
     */
    public abstract int getPoints();
    
    /**
     * Get the alien type
     * 
     * @return Alien type
     */
    public abstract Type getType();
    
    
    @Override
    public void update() {
        super.update();
    }
    
    @Override
    public void dump() {
        System.out.println("Alien:");
        System.out.println("  Name: " + this.name);
        System.out.println("  Type: " + getType());
        System.out.println("  Position: (" + this.x + ", " + this.y + ")");
        System.out.println("  Points: " + getPoints());
    }
}